csharp/src/Apache.Arrow/C/CArrowArrayStreamImporter.cs

```diff
@@ -0,0 +1,137 @@
+        /// </summary>
+        /// <remarks>
+        /// This will call the release callback on the passed struct if the function fails.
+        /// Otherwise, the release callback is called when the IArrowArrayStream is disposed.
+        /// </remarks>
+        /// <examples>
+        /// Typically, you will allocate an uninitialized CArrowArrayStream pointer,
+        /// pass that to external function, and then use this method to import
+        /// the result.
+        /// 
+        /// <code>
+        /// CArrowArrayStream* importedPtr = CArrowArrayStream.Create();
+        /// foreign_export_function(importedPtr);
+        /// IArrowArrayStream importedStream = CArrowArrayStreamImporter.ImportStream(importedPtr);
+        /// </code>
+        /// </examples>
+        public static unsafe IArrowArrayStream ImportArrayStream(CArrowArrayStream* ptr)
+        {
+            return new ImportedArrowArrayStream(ptr);
+        }
+
+        private sealed unsafe class ImportedArrowArrayStream : IArrowArrayStream
+        {
+            private readonly CArrowArrayStream* _cArrayStream;
+            private readonly Schema _schema;
+            private bool _disposed;
+
+            public ImportedArrowArrayStream(CArrowArrayStream* cArrayStream)
+            {
+                if (cArrayStream == null)
+                {
+                    throw new ArgumentNullException(nameof(cArrayStream));
+                }
+                _cArrayStream = cArrayStream;
+                if (_cArrayStream->release == null)
+                {
+                    throw new ArgumentException("Tried to import an array stream that has already been released.", nameof(cArrayStream));
+                }
+
+                CArrowSchema* cSchema = CArrowSchema.Create();
+                try
+                {
+                    int errno = _cArrayStream->get_schema(_cArrayStream, cSchema);
+                    if (errno != 0)
+                    {
+                        throw new Exception($"Unexpected error recieved from external stream. Errno: {errno}");
+                    }
+                    _schema = CArrowSchemaImporter.ImportSchema(cSchema);
+                }
+                finally
+                {
+                    if (_schema == null)
+                    {
+                        CArrowSchema.Free(cSchema);
+                    }
+                }
+            }
+
+            ~ImportedArrowArrayStream()
+            {
+                Dispose();
+            }
+
+            public Schema Schema => _schema;
+
+            public ValueTask<RecordBatch> ReadNextRecordBatchAsync(CancellationToken cancellationToken = default)
+            {
+                if (_disposed)
+                {
+                    throw new ObjectDisposedException(typeof(ImportedArrowArrayStream).Name);
+                }
+
+                RecordBatch result = null;
+                CArrowArray* cArray = CArrowArray.Create();
```

Comment:
Hi @CurtHagenlocher, thanks for implementing this! I've started testing using this to support reading Parquet data as Arrow record batches in [ParquetSharp](https://github.com/G-Research/ParquetSharp).

One concern I've found is that it appears that an imported stream will leak memory as these `CArrowArray` instances are allocated for each batch but they're never freed, as `ImportedArrowArray.FinalRelease` will call the `release` callback but never deallocates the `CArrowArray` struct itself.

Is that correct or am I missing something here?

Would it make sense for the imported array to take ownership of the `CArrowArray` struct and deallocate it after calling release?

This seems to be an issue for external use of the C Data Interface API too, eg. if I want to return an `IArrowArrayStream` from a method I can't free the `CArrowArrayStream` that was used to import it until after the user is finished with the stream, which is awkward.

Reply:
Yes, this has already been discovered and there's a PR out for fixing it: https://github.com/apache/arrow/pull/35810

Reply:
Okay, my pattern matching was a bit too eager for the first problem. Yes, that looks like a leak and I think your suggestion is right; that on this code path the ImportedArrowArray needs to remember that it owns the allocation.

The second problem feels different, because in the first case we always know that we were the ones who allocated the CArrowArray but I'm not entirely sure we know that about the CArrowArrayStream. It's true for the stream we got from Python in the test case, but that was using a pyarrow-specific API. The flavor of the C API as a whole does suggest that it will usually be the case for the caller to have to allocate the CArrowArrayStream without (I think) quite making it explicit.

On the whole, I suspect that both importers should take a flag which says whether or not to deallocate the structure afterwards. I'm not convinced about the right default for the flag given the relative risks of leaking memory vs deallocating it inappopriately.

Reply:
Filed #35988

Reply:
Right, yeah having a flag to say whether the structure should be deallocated would solve both problems if it was public and implemented for both the array and stream. I think keeping the current behaviour of not deallocating makes sense as a default, given a small memory leak is less of an issue than incorrectly deallocating, and then that keeps allocation and deallocation symmetric by default.

It wouldn't be a big deal if only the first problem was solved though, I can solve the second problem with some extra indirection by introducing a wrapper for the imported stream that deallocates the struct on dispose.

Reply:
https://github.com/apache/arrow/pull/35996

Reply:
Thanks! That solution of moving the struct into the imported array or stream is much nicer